src/machines/activity_state_machine.rs

```diff
@@ -262,32 +272,126 @@ impl Created {
 pub(super) struct ScheduleCommandCreated {}
 
 impl ScheduleCommandCreated {
-    pub(super) fn on_activity_task_scheduled(self) -> ActivityMachineTransition {
-        // set initial command event id
-        //  this.initialCommandEventId = currentEvent.getEventId();
+    pub(super) fn on_activity_task_scheduled(
+        self,
+        mut dat: SharedState,
+        scheduled_event_id: i64,
+    ) -> ActivityMachineTransition {
+        dat.scheduled_event_id = scheduled_event_id;
         ActivityMachineTransition::default::<ScheduledEventRecorded>()
     }
-    pub(super) fn on_canceled(self) -> ActivityMachineTransition {
-        // cancelCommandNotifyCanceled
-        ActivityMachineTransition::default::<Canceled>()
+    pub(super) fn on_canceled(self, dat: SharedState) -> ActivityMachineTransition {
+        let canceled_state = SharedState {
+            cancelled_before_sent: true,
+            ..dat
+        };
+        match dat.cancellation_type {
+            ActivityCancellationType::Abandon => ActivityMachineTransition::ok_shared(
+                vec![ActivityMachineCommand::Cancel],
+                Canceled::default(),
+                canceled_state,
+            ),
+            _ => Common::notify_canceled(canceled_state, Canceled::default().into()),
+        }
     }
 }
 
 #[derive(Default, Clone)]
 pub(super) struct ScheduledEventRecorded {}
 
 impl ScheduledEventRecorded {
-    pub(super) fn on_task_started(self) -> ActivityMachineTransition {
-        // setStartedCommandEventId
+    pub(super) fn on_task_started(
+        self,
+        mut dat: SharedState,
+        started_event_id: i64,
+    ) -> ActivityMachineTransition {
+        dat.started_event_id = started_event_id;
         ActivityMachineTransition::default::<Started>()
     }
     pub(super) fn on_task_timed_out(self) -> ActivityMachineTransition {
         // notify_timed_out
         ActivityMachineTransition::default::<TimedOut>()
     }
-    pub(super) fn on_canceled(self) -> ActivityMachineTransition {
+    pub(super) fn on_canceled(self, dat: SharedState) -> ActivityMachineTransition {
+        Common::create_request_cancel_activity_task_command(dat)
+    }
+}
+
+struct Common {}
+impl Common {
```

Comment:
I don't think this `Common` struct serves any purpose. You can implement these as freestanding functions.

Reply:
I thought of it as just a way to aggregate helpers together, doesn't need to be that way though, moved functions to the top.